dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/reservedvalue/RandomGeneratorService.java

```diff
@@ -0,0 +1,207 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.reservedvalue;
+
+import static org.hisp.dhis.util.Constants.RANDOM_GENERATION_CHUNK;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import lombok.AccessLevel;
+import lombok.Setter;
+
+import org.springframework.stereotype.Service;
+
+/**
+ * @author Luca Cambi <luca@dhis2.org>
+ *
+ *         Generate random values from a pattern using 256 bit BigInteger for
+ *         numbers and random UUID for alphanumerics.
+ *
+ *         x = lower case
+ *
+ *         X = upper case
+ *
+ *         # = digit
+ *
+ *         * = digit or lower case or upper case
+ *
+ *         For alphanumeric we have some corner cases:
+ *
+ *         - '*' it shuffles between lower and upper case and gets a number when
+ *         if finds it in UUID random string
+ *
+ *         - 'x' or 'X' if we find a number in UUID random string it gets the
+ *         alphabetic char at that position in order to maintain random pattern
+ *
+ */
+@Service
+@Setter( AccessLevel.PROTECTED )
+public class RandomGeneratorService implements Callable<List<String>>
+{
+
+    private final List<Character> lowercase = IntStream.range( 0, 26 ).mapToObj( n -> (char) (n + 'a') )
+        .collect( Collectors.toList() );
+
+    private final List<Character> uppercase = IntStream.range( 0, 26 ).mapToObj( n -> (char) (n + 'A') )
+        .collect( Collectors.toList() );
+
+    private String textPattern;
+
+    @Override
+    public List<String> call()
+        throws Exception
+    {
+        LinkedList<String> patterns = new LinkedList<>();
+
+        List<String> randomList = new ArrayList<>();
+
+        Pattern randomPattern = Pattern.compile( "[X]+|[x]+|[#]+|[*]+" );
+        Matcher matcher = randomPattern.matcher( textPattern );
+
+        while ( matcher.find() )
+        {
+            patterns.add( textPattern.substring( matcher.start(), matcher.end() ) );
+        }
+
+        for ( int j = 0; j < RANDOM_GENERATION_CHUNK; j++ )
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+
+            for ( String pattern : patterns )
+            {
+                switch ( pattern.charAt( 0 ) )
+                {
+                case '*':
+                    setForRandomAll( stringBuilder, pattern );
+                    break;
+                case '#':
+                    stringBuilder.append( new BigInteger( 256, new SecureRandom() ).abs().toString(), 0,
```

Comment:
Just asking. Do we really need a new Random every time here ? Can't the same instance be reused ?

Reply:
yes good point i think we can reuse the same, allthough ideally i think just one would be used in case of a chunck of integers